test/src/main/java/org/springframework/security/test/support/JwtAuthenticationBuilder.java

```diff
@@ -0,0 +1,227 @@
+		this.authorities.clear();
+		authorities.forEach(this::authority);
+		return downCast();
+	}
+
+	public T authorities(final String... authorities) {
+		return authorities(Stream.of(authorities));
+	}
+
+	public T authorities(final Collection<String> authorities) {
+		return authorities(authorities.stream());
+	}
+
+	public T role(final String role) {
+		Assert.hasText(role, "authority must be non empty");
+		Assert.isTrue(
+				!role.startsWith(ROLE_PREFIX),
+				"role must not be prefixed with " + ROLE_PREFIX + " (it is auto-added)");
+		return authority(ROLE_PREFIX + role);
+	}
+
+	public T roles(final Stream<String> roles) {
+		this.authorities.removeIf(a -> a.startsWith(ROLE_PREFIX));
+		roles.forEach(this::role);
+		return downCast();
+	}
+
+	public T roles(final String... roles) {
+		return roles(Stream.of(roles));
+	}
+
+	public T roles(final Collection<String> roles) {
+		return roles(roles.stream());
+	}
+
+	public T claim(final String name, final Object value) {
+		Assert.hasText(name, "Claim name must be non empty");
+		if (getNameClaimName().equals(name)) {
+			return name(value.toString());
+		}
+		this.claims.put(name, value);
+		return downCast();
+	}
+
+	public T claims(final Map<String, Object> attributes) {
+		Assert.notNull(attributes, "attributes must be non null");
+		this.claims.clear();
+		attributes.entrySet().stream().forEach(e -> this.claim(e.getKey(), e.getValue()));
+		return downCast();
+	}
+
+	public T authoritiesConverter(final Converter<Jwt, Collection<GrantedAuthority>> authoritiesConverter) {
+		this.authoritiesConverter = authoritiesConverter;
+		return downCast();
+	}
+
+	Set<GrantedAuthority> getAllAuthorities(final Jwt token) {
+		return Stream
+				.concat(
+						authorities.stream().map(SimpleGrantedAuthority::new),
+						authoritiesConverter.convert(token).stream())
+				.collect(Collectors.toSet());
+	}
+
+	public T header(final String name, final Object value) {
+		Assert.hasText(name, "Header name can't be empty");
+		this.headers.put(name, value);
+		return downCast();
+	}
+
+	public T headers(final Map<String, Object> headers) {
+		Assert.notEmpty(headers, "Headers can't be empty as Jwt constructor throws runtime exception on empty headers.");
+		this.headers.clear();
+		headers.entrySet().stream().forEach(e -> this.header(e.getKey(), e.getValue()));
+		return downCast();
+	}
+
+	public JwtAuthenticationToken build() {
+		putIfNotEmpty(getNameClaimName(), name, claims);
+
+		final Jwt token = new Jwt(
+				tokenValue,
+				(Instant) claims.get(JwtClaimNames.IAT),
+				(Instant) claims.get(JwtClaimNames.EXP),
+				headers.isEmpty() ? DEFAULT_HEADERS : headers,
+				claims);
+
+		return new JwtAuthenticationToken(token, getAllAuthorities(token));
+	}
+
+	protected String getNameClaimName() {
+		return JwtClaimNames.SUB;
+	}
+
+	@SuppressWarnings("unchecked")
+	protected T downCast() {
+		return (T) this;
+	}
+
+	private static final Map<String, Object>
```

Comment:
Not sure :/

I tried to open a [ticket regarding potential incoherence in AbstractOAuth2Token](https://github.com/spring-projects/spring-security/issues/6807) but it was closed "won't fix".

How to decide the order in witch we should process `token.getIssuedAt()`, `token.getExpiresAt()` and `claims` (which main contain `JwtClaimNames.IAT` and `JwtClaimNames.EXP`) for the "last one wins" to be applicable?

As a side note, I understand you don't want to introduce breaking changes at this stage. Few remarks on the working branch I linked to this ticket:
  * neither marking `getIssuedAt()` and `getExpiresAt()` as abstract (and providing implementations retrieving it from attributes in concreate classes) nor removing `issuedAt` and `expiresAt` properties in `AbstractOAuth2Token` requires breaking changes
  * I don't intend to author anything concerning core objects. To me this is team responsibility: I have neither skills nor background for that. This branch is just a support to explore possibilities and get an idea of impacts. Don't expect atomic changes on it, again it is not intended to be integrated in your codebase.
  * breaking changes on this branch due other potential changes orthogonal to this concern:
    - promoting `attributes` as `AbstractOAuth2Token` as all OAuth2 tokens types have some (even refresh-tokens are issued and might expire)
    - making `AbstractOAuth2Token` a `Principal`. I was first surprised that `principal` property in `AbstractOAuth2TokenAuthenticationToken` is neither a `java.security.Principal` nor an `AuthenticatedPrincipal`. This change allows to change that and quite a few things suddenly get clearer to me

Reply:
In this case, it might be simpler to show what I'm thinking as I agree with you that the phrase "last one wins" might not be the best way to put it. I think a builder should understand that some opinions are stronger than others, so, for example, calling `claims(Map)` should take precedence over and `claim(name, value)` calls. But, several calls to `claim(name, value)`, the last one should win (in case of a name conflict). In circumstances where the strength of the user's opinion can't be differentiated (maybe `authorities` and `roles` is an example of this), then whichever was called last should win out.

Anyway, code probably speaks more clearly than I can articulate in words. :) This is what I'm thinking might be an option: https://github.com/jzheaux/spring-security/commit/44192fdb4890ae686d7a0a7ffe70ab98219516d8 Warning, it's a rough cut, but I'm just trying to articulate the idea,